File: fabric/example-mod/src/main/java/com/example/ExampleMod.java
package com.example;

import dev.faststats.core.ErrorTracker;
import dev.faststats.core.Metrics;
import dev.faststats.core.Settings;
import dev.faststats.core.data.Metric;
import dev.faststats.fabric.FabricMetrics;
import net.fabricmc.api.ModInitializer;

public class ExampleMod implements ModInitializer {
    private final Metrics metrics = FabricMetrics.factory()
            // Custom metrics require a corresponding data source in your project settings
            .addMetric(Metric.number("example_metric", () -> 42))

            // Error tracking must be enabled in the project settings
            .errorTracker(ErrorTracker.contextAware())

            .settings(Settings.withToken("YOUR_TOKEN_HERE")) // token can be found in the settings of your project
            .create("example-mod"); // your mod id as defined in fabric.mod.json

    @Override
    public void onInitialize() {
    }
}

File: fabric/example-mod/src/main/java/com/example/ExampleMod.kt
package com.example

import dev.faststats.ErrorTracker
import dev.faststats.data.Metric
import dev.faststats.fabric.FabricContext
import net.fabricmc.api.ModInitializer
import java.util.concurrent.atomic.AtomicInteger

class ExampleMod : ModInitializer {
    private val gameCount = AtomicInteger()

    private val context = FabricContext.Factory(
        "example-mod", // your mod id as defined in fabric.mod.json
        "YOUR_TOKEN_HERE",
    )
        // .metrics(Metrics.Factory::create) // Define a minimal metrics instance without any custom metrics
        .metrics { factory ->
            factory
                // Custom metrics require a corresponding data source in your project settings
                .addMetric(Metric.number("game_count") { gameCount.get() })
                .addMetric(Metric.string("server_version") { "1.0.0" })

                // #onFlush is invoked after successful metrics submission
                // This is useful for cleaning up cached data
                .onFlush { gameCount.set(0) } // reset game count on flush

                .create()
        }
        .errorTrackerService(ERROR_TRACKER)
        .create()

    override fun onInitialize() {
        // your actual logic
    }

    fun startGame() {
        gameCount.incrementAndGet()
    }

    companion object {
        val ERROR_TRACKER: ErrorTracker = ErrorTracker.contextAware()
    }
}
